lib/compat.php

```diff
@@ -367,91 +367,90 @@ function gutenberg_replace_default_block_categories( $default_categories ) {
-		$context['query'] = array( 'categoryIds' => array() );
+function gutenberg_render_block_with_assigned_block_context( $args ) {
+	if ( is_callable( $args['render_callback'] ) ) {
+		$block_render_callback   = $args['render_callback'];
+		$args['render_callback'] = function( $attributes, $content, $block = null ) use ( $block_render_callback ) {
+			global $post, $wp_query;
+
+			// Check for null for back compatibility with WP_Block_Type->render
+			// which is unused since the introduction of WP_Block class.
+			//
+			// See:
+			// - https://core.trac.wordpress.org/ticket/49927
+			// - commit 910de8f6890c87f93359c6f2edc6c27b9a3f3292 at wordpress-develop.
+
+			if ( null === $block ) {
+				return $block_render_callback( $attributes, $content );
+			}
 
-		foreach ( $wp_query->tax_query->queried_terms['category']['terms'] as $category_slug_or_id ) {
-			$context['query']['categoryIds'][] = 'slug' === $wp_query->tax_query->queried_terms['category']['field'] ? get_cat_ID( $category_slug_or_id ) : $category_slug_or_id;
-		}
-	}
+			if ( property_exists( $block, 'context' ) ) {
+				$context = array();
 
-	if ( isset( $wp_query->tax_query->queried_terms['post_tag'] ) ) {
-		if ( isset( $context['query'] ) ) {
-			$context['query']['tagIds'] = array();
-		} else {
-			$context['query'] = array( 'tagIds' => array() );
-		}
+				if ( $post instanceof WP_Post ) {
+					$context['postId'] = $post->ID;
 
-		foreach ( $wp_query->tax_query->queried_terms['post_tag']['terms'] as $tag_slug_or_id ) {
-			$tag_ID = $tag_slug_or_id;
+					/*
+					* The `postType` context is largely unnecessary server-side, since the
+					* ID is usually sufficient on its own. That being said, since a block's
+					* manifest is expected to be shared between the server and the client,
+					* it should be included to consistently fulfill the expectation.
+					*/
+					$context['postType'] = $post->post_type;
+				}
 
-			if ( 'slug' === $wp_query->tax_query->queried_terms['post_tag']['field'] ) {
-				$tag = get_term_by( 'slug', $tag_slug_or_id, 'post_tag' );
+				if ( isset( $wp_query->tax_query->queried_terms['category'] ) ) {
+					$context['query'] = array( 'categoryIds' => array() );
 
-				if ( $tag ) {
-					$tag_ID = $tag->term_id;
+					foreach ( $wp_query->tax_query->queried_terms['category']['terms'] as $category_slug_or_id ) {
+						$context['query']['categoryIds'][] = 'slug' === $wp_query->tax_query->queried_terms['category']['field'] ? get_cat_ID( $category_slug_or_id ) : $category_slug_or_id;
+					}
 				}
-			}
-			$context['query']['tagIds'][] = $tag_ID;
-		}
-	}
 
-	/**
-	 * Filters the default context provided to a rendered block.
-	 *
-	 * @param array $context      Default context.
-	 * @param array $parsed_block Block being rendered, filtered by `render_block_data`.
-	 */
-	$context = apply_filters( 'render_block_context', $context, $parsed_block );
+				if ( isset( $wp_query->tax_query->queried_terms['post_tag'] ) ) {
```

Comment:
@ntsekouras I'm not sure I  understand why we need to inejct the "query"  context here. This compact function should be only used to implement the context support behavior and not inject things into the context. unless,  these are "default injections" things that should be injected to all blocks?

Reply:
The first injection for `categories` was from Enrique and when I implemented the `tags` support followed that lead, without digging deep. After some more recent explorations with context (without being sure yet), I believe this was implemented by Enrique to make the `QueryLoop` inherit global context if is not wrapped by `Query` block.

I believe this used to work for the above case (not wrapped in Query), but many changes to many things (like FSE navigation, context, Query etc) made it not to use these values. 

For me this code is a part to the exploration of inheriting global context to Query block, but it's not clear to me how to handle this.

Reply:
@ntsekouras Would you be able to check that everything works properly for the query block in this PR?

Reply:
Query works fine, but I said above, I think this is some leftover code that needs reworking as a part of global context inheritance problem.

Reply:
yeah, that's fine, I'm just making sure I don't regress anything and fix the currently failing e2e tests. Thanks for testing.